synapse/config/server.py

```diff
@@ -228,6 +228,90 @@ def read_config(self, config):
             "show_users_in_user_directory", True,
         )
 
+        retention_config = config.get("retention", {})
+
+        self.retention_enabled = retention_config.get("enabled", False)
+
+        self.retention_min_lifetime = self.parse_duration(
+            retention_config.get("min_lifetime", 0)
+        )
+
+        self.retention_max_lifetime = self.parse_duration(
+            retention_config.get("max_lifetime", 0)
+        )
+
+        if self.retention_enabled and not (
+            self.retention_min_lifetime and self.retention_max_lifetime
+        ):
+            raise ConfigError(
+                "'min_lifetime' and 'max_lifetime' must be provided if the message"
```

Comment:
why, given it's a cap? why not make it so that there can be no cap?

Reply:
As the comment states, most of the code currently relies on there being caps (mostly because it's part of the initial feature request) and I don't believe there's time for me to both remove this limitation and land the feature before the deadline.

I agree it's not ideal, and it's mainly my fault for not thinking it through enough, and it's on my TODO list to iterate over this PR and make the implem more generic by removing this constraint (among other things).